ndc-lambda-sdk/src/inference.ts

```diff
@@ -422,38 +488,69 @@ function isJSONValueType(tsType: ts.Type, ndcLambdaSdkModule: ts.ResolvedModuleF
+function deriveSchemaTypeIfObjectType(tsType: ts.Type, typePath: schema.TypePathSegment[], allowRelaxedTypes: boolean, context: TypeDerivationContext, recursionDepth: number): Result<schema.TypeReference, string[]> | undefined {
   const info = getObjectTypeInfo(tsType, typePath, context.typeChecker, context.functionsFilePath);
   if (info) {
+    const makeRelaxedTypesError : () => Result<schema.TypeReference, string[]> = () => new Err<schema.TypeReference, string[]>([`The object type '${info.generatedTypeName}' uses relaxed types and can only be used by a function marked with @allowrelaxedtypes. It was encountered in ${schema.typePathToString(typePath)}`]);
+
     // Short-circuit recursion if the type has already been named
-    if (context.objectTypeDefinitions[info.generatedTypeName]) {
+    const existingType = context.objectTypeDefinitions[info.generatedTypeName];
+    if (existingType) {
+      if (allowRelaxedTypes === false && existingType.isRelaxedType)
+        return makeRelaxedTypesError();
+
       return new Ok({ type: 'named', name: info.generatedTypeName, kind: "object" });
     }
 
-    context.objectTypeDefinitions[info.generatedTypeName] = { properties: [], description: null }; // Break infinite recursion
+    context.objectTypeDefinitions[info.generatedTypeName] = { properties: [], description: null, isRelaxedType: false }; // Break infinite recursion
 
-    const propertyResults = Result.traverseAndCollectErrors(Array.from(info.properties), ([propertyName, propertyInfo]) => {
-      return deriveSchemaTypeForTsType(propertyInfo.tsType, [...typePath, { segmentType: "ObjectProperty", typeName: info.generatedTypeName, propertyName }], context, recursionDepth + 1)
+    return Result.traverseAndCollectErrors(Array.from(info.properties), ([propertyName, propertyInfo]) => {
+      return deriveSchemaTypeForTsType(propertyInfo.tsType, [...typePath, { segmentType: "ObjectProperty", typeName: info.generatedTypeName, propertyName }], allowRelaxedTypes, context, recursionDepth + 1)
         .map(propertyType => ({ propertyName: propertyName, type: propertyType, description: propertyInfo.description }));
+    })
+    .bind(propertyResults => {
+      const isRelaxedType = propertyResults.some(propDef => isTypeReferenceARelaxedType(propDef.type, context));
+      if (allowRelaxedTypes === false && isRelaxedType) {
+        return makeRelaxedTypesError();
+      }
+
+      context.objectTypeDefinitions[info.generatedTypeName] = { properties: propertyResults, description: info.description, isRelaxedType }
+      return new Ok({ type: 'named', name: info.generatedTypeName, kind: "object" } as const)
+    })
+    .onErr(_err => {
+      delete context.objectTypeDefinitions[info.generatedTypeName]; // Remove the recursion short-circuit to allow other functions to try making this type again
     });
+  }
+}
```

Comment:
Quite a lot of changes, so a bit hard to tell if anything is off. I'll focus on the test-cases and see if that all looks to preserve the behaviour.